src/raftstore/store/peer.rs

```diff
@@ -1241,14 +1237,115 @@ impl Peer {
 
-    fn handle_raft_committed_entries(&mut self,
-                                     committed_entries: Vec<eraftpb::Entry>)
-                                     -> Vec<ExecResult> {
+    fn exec_read(&mut self, req: &RaftCmdRequest) -> Result<RaftCmdResponse> {
+        let snap = Snapshot::new(self.engine.clone());
+        let requests = req.get_requests();
+        let mut responses = Vec::with_capacity(requests.len());
+
+        for req in requests {
+            let cmd_type = req.get_cmd_type();
+            let mut resp = match cmd_type {
+                CmdType::Get => try!(do_get(&self.tag, self.region(), &snap, req)),
+                CmdType::Snap => try!(do_snap(self.region().to_owned())),
+                CmdType::Put | CmdType::Delete | CmdType::Invalid => unreachable!(),
+            };
+
+            resp.set_cmd_type(cmd_type);
+
+            responses.push(resp);
+        }
+
+        let mut resp = RaftCmdResponse::new();
+        resp.set_responses(protobuf::RepeatedField::from_vec(responses));
+        Ok(resp)
+    }
+}
+
+fn check_data_key(key: &[u8], region: &metapb::Region) -> Result<()> {
+    // region key range has no data prefix, so we must use origin key to check.
+    try!(util::check_key_in_region(key, region));
+
+    Ok(())
+}
+
+fn do_get(tag: &str, region: &metapb::Region, snap: &Snapshot, req: &Request) -> Result<Response> {
+    // TODO: the get_get looks wried, maybe we should figure out a better name later.
+    let key = req.get_get().get_key();
+    try!(check_data_key(key, &region));
+
+    let mut resp = Response::new();
+    let res = if req.get_get().has_cf() {
+        let cf = req.get_get().get_cf();
+        // TODO: check whether cf exists or not.
+        snap.get_value_cf(cf, &keys::data_key(key)).unwrap_or_else(|e| {
+            panic!("{} failed to get {} with cf {}: {:?}",
+                   tag,
+                   escape(key),
+                   cf,
+                   e)
+        })
+    } else {
+        snap.get_value(&keys::data_key(key))
+            .unwrap_or_else(|e| panic!("{} failed to get {}: {:?}", tag, escape(key), e))
+    };
+    if let Some(res) = res {
+        resp.mut_get().set_value(res.to_vec());
+    }
+
+    Ok(resp)
+}
+
+fn do_snap(region: metapb::Region) -> Result<Response> {
+    let mut resp = Response::new();
+    resp.mut_snap().set_region(region);
+    Ok(resp)
+}
+
+pub struct ApplyDelegate {
+    // peer_id
+    pub id: u64,
+    // peer_tag, "[region region_id] peer_id"
+    pub tag: String,
+    pub engine: Arc<DB>,
+    pub region: metapb::Region,
+    // if we remove ourself in ChangePeer remove, we should set this flag, then
+    // any following committed logs in same Ready should be applied failed.
+    pub pending_remove: bool,
+    pub apply_state: RaftApplyState,
+    pub applied_index_term: u64,
+    pub term: u64,
+    pub pending_cmds: PendingCmdQueue,
+    pub metrics: ApplyMetrics,
+}
+
+impl ApplyDelegate {
+    pub fn new(id: u64,
```

Comment:
IMO, there is no big difference using options or not. Because I have to pass these much args to construct an option too.